# Tarea 2 -  Regresión Lineal Múltiple 

### ITESM - Escuela de Ingeniería y Ciencias
#### Maestría en Ciencias Computacionales
#### Aprendizaje Automático

---

## Carlos E. Hdz Rincón - A01181616

#1. Carga de datos
En este ejercicio usaremos la base de datos de Boston_housing_dataset
(https://www.kaggle.com/apratim87/housingdata) la cual consta de 13 variables
independientes y la variable dependiente. Son 506 datos de casas cuyo
objetivo es la predicción de su costo (MEDV).

In [ ]:
import numpy as np    # biblioteca para el manejo de vectores y matrices en python.
import pandas as pd   # biblioteca para el manejo de datos y tablas.
import matplotlib.pyplot as plt   # biblioteca para visualización de datos.
import statsmodels.api as sm      # biblioteca para pruebas estadísticas.

# Connect to Google Drive
#from google.colab import drive
#drive.mount('/content/drive')

Housing data variables in order:

1. CRIM per capita crime rate by town
1. ZN proportion of residential land zoned for lots over 25,000 sq.ft.
1. INDUS proportion of non-retail business acres per town
1. CHAS Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
1. NOX nitric oxides concentration (parts per 10 million)
1. RM average number of rooms per dwelling
1. AGE proportion of owner-occupied units built prior to 1940
1. DIS weighted distances to five Boston employment centres
1. RAD index of accessibility to radial highways
1. TAX full-value property-tax rate per \$10,000
1. PTRATIO pupil-teacher ratio by town
1. B $1000(Bk - 0.63)^2$ where Bk is the proportion of blacks by town
1. LSTAT % lower status of the population
1. MEDV Median value of owner-occupied homes in $1000's

In [ ]:
housing_data_location = "/content/drive/My Drive/Datasets/BostonHousingDataset/housingdata.csv"
headers = ["CRIM","ZN","INDUS","CHAS","NOX","RM","AGE","DIS","RAD","TAX",
           "PTRATIO","B","LSTAT","MEDV"]
mydata = pd.read_csv(housing_data_location, names=headers)
mydata.head()

CRIM    ZN  INDUS  CHAS    NOX  ...  TAX  PTRATIO       B  LSTAT  MEDV
0  0.00632  18.0   2.31     0  0.538  ...  296     15.3  396.90   4.98  24.0
1  0.02731   0.0   7.07     0  0.469  ...  242     17.8  396.90   9.14  21.6
2  0.02729   0.0   7.07     0  0.469  ...  242     17.8  392.83   4.03  34.7
3  0.03237   0.0   2.18     0  0.458  ...  222     18.7  394.63   2.94  33.4
4  0.06905   0.0   2.18     0  0.458  ...  222     18.7  396.90   5.33  36.2

[5 rows x 14 columns]

In [ ]:
print(mydata.shape)

(506, 14)


#2. Obtener el modelo de Regresión Lineal Múltiple (RLM) del modelo
Con base a dicho modelo contestar los siguientes incisos.

## Preparación de datos

Separamos las variables dependientes de la variable independiente MEDV que es el valor medio de las casas ocupadas en 1000s de USD.

In [ ]:
# Separamos las variables independientes de la dependiente MEDV que es el valor
x = mydata.loc[:, mydata.columns != "MEDV"]
print(x.shape)
print(x.head())

y = mydata[["MEDV"]]
print(y.shape)
print(y.head())

(506, 13)
      CRIM    ZN  INDUS  CHAS    NOX  ...  RAD  TAX  PTRATIO       B  LSTAT
0  0.00632  18.0   2.31     0  0.538  ...    1  296     15.3  396.90   4.98
1  0.02731   0.0   7.07     0  0.469  ...    2  242     17.8  396.90   9.14
2  0.02729   0.0   7.07     0  0.469  ...    2  242     17.8  392.83   4.03
3  0.03237   0.0   2.18     0  0.458  ...    3  222     18.7  394.63   2.94
4  0.06905   0.0   2.18     0  0.458  ...    3  222     18.7  396.90   5.33

[5 rows x 13 columns]
(506, 1)
   MEDV
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2


In [ ]:
print(x)


        CRIM    ZN  INDUS  CHAS    NOX  ...  RAD  TAX  PTRATIO       B  LSTAT
0    0.00632  18.0   2.31     0  0.538  ...    1  296     15.3  396.90   4.98
1    0.02731   0.0   7.07     0  0.469  ...    2  242     17.8  396.90   9.14
2    0.02729   0.0   7.07     0  0.469  ...    2  242     17.8  392.83   4.03
3    0.03237   0.0   2.18     0  0.458  ...    3  222     18.7  394.63   2.94
4    0.06905   0.0   2.18     0  0.458  ...    3  222     18.7  396.90   5.33
..       ...   ...    ...   ...    ...  ...  ...  ...      ...     ...    ...
501  0.06263   0.0  11.93     0  0.573  ...    1  273     21.0  391.99   9.67
502  0.04527   0.0  11.93     0  0.573  ...    1  273     21.0  396.90   9.08
503  0.06076   0.0  11.93     0  0.573  ...    1  273     21.0  396.90   5.64
504  0.10959   0.0  11.93     0  0.573  ...    1  273     21.0  393.45   6.48
505  0.04741   0.0  11.93     0  0.573  ...    1  273     21.0  396.90   7.88

[506 rows x 13 columns]


In [ ]:
xones = sm.add_constant(x)  # Agregamos una columna de 1s
print(xones.head())

   const     CRIM    ZN  INDUS  CHAS  ...  RAD  TAX  PTRATIO       B  LSTAT
0    1.0  0.00632  18.0   2.31     0  ...    1  296     15.3  396.90   4.98
1    1.0  0.02731   0.0   7.07     0  ...    2  242     17.8  396.90   9.14
2    1.0  0.02729   0.0   7.07     0  ...    2  242     17.8  392.83   4.03
3    1.0  0.03237   0.0   2.18     0  ...    3  222     18.7  394.63   2.94
4    1.0  0.06905   0.0   2.18     0  ...    3  222     18.7  396.90   5.33

[5 rows x 14 columns]


## Modelo de regresion lineal multivariable

In [ ]:
model_all_columns = sm.OLS(y, xones).fit()
print(model_all_columns.summary())

                            OLS Regression Results                            
Dep. Variable:                   MEDV   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     108.1
Date:                Tue, 25 Aug 2020   Prob (F-statistic):          6.72e-135
Time:                        22:27:45   Log-Likelihood:                -1498.8
No. Observations:                 506   AIC:                             3026.
Df Residuals:                     492   BIC:                             3085.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         36.4595      5.103      7.144      0.0

##i. Interpretar el valor del coeficiente de determinación ajustado $𝑅^2$

Con un valor $\bar{R^2}=0.734$ podemos apreciar que se tiene una correlación importante entre las variables independientes y la variable dependiente, explicándose con un 73.4% de la variabilidad de la variable dependiente queda explicada con las 13 variables independientes utilizadas.


##ii. Determina qué variables no resultaron significativas considerendo un 𝑣𝑎𝑙𝑜𝑟-𝑝 = 0.01.

En el reporte del modelo se puede apreciar que INDUS y AGE poseen valores p de 0.738 y 0.958 respectivamente, que nos indican que probablemente no están aportando mucho valor a la predicción de la variable dependiente. 


##iii. Obtener un segundo modelo de RLM omitiendo las variables que no resultaron significativas en el inciso anterior.

Comparar su coeficiente de determinación ajustado con base con respecto al primer modelo.

In [ ]:
# Removemos las columnas INDUS y AGE
x2 = x.loc[:, x.columns.difference(['INDUS', 'AGE'])]
print(x2.shape)
print(x2.head())

xones2 = sm.add_constant(x2)  # Agregamos una columna de 1s

model2 = sm.OLS(y, xones2).fit()
print(model2.summary())

(506, 11)
        B  CHAS     CRIM     DIS  LSTAT    NOX  PTRATIO  RAD     RM  TAX    ZN
0  396.90     0  0.00632  4.0900   4.98  0.538     15.3    1  6.575  296  18.0
1  396.90     0  0.02731  4.9671   9.14  0.469     17.8    2  6.421  242   0.0
2  392.83     0  0.02729  4.9671   4.03  0.469     17.8    2  7.185  242   0.0
3  394.63     0  0.03237  6.0622   2.94  0.458     18.7    3  6.998  222   0.0
4  396.90     0  0.06905  6.0622   5.33  0.458     18.7    3  7.147  222   0.0
                            OLS Regression Results                            
Dep. Variable:                   MEDV   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.735
Method:                 Least Squares   F-statistic:                     128.2
Date:                Tue, 25 Aug 2020   Prob (F-statistic):          5.54e-137
Time:                        22:27:53   Log-Likelihood:                -1498.9
No. Observations:                 506   AI

---

Después de remover las columnas INDUS y AGE y con un valor $\bar{R^2}=0.735$ podemos apreciar que hay una ganancia de 0.001 que parece no ser mucho, sin embargo es significativo ya que comprueba la suposición de que no estaban aportando mucho al modelo. Aún más interesante, los valores P de las variables restantes se ven bastante bien al acercarse a 0, lo que indica que los coeficientes son los mejores que se pueden obtener.

Este modelo tiene el potencial de ser el mejor que se puede obtener para este problema.

##iv. Obtener la matriz de correlación de Pearson de las 14 variables.
Con base a dicha matriz identifica los pares de variables independientes que tengan un coeficiente de correlación mayor o igual a 0.7 en valor absoluto. Obtener un tercer modelo de RLM y compara la significacia de los coeficientes y su coeficiente de determinación ajustado con respecto a los dos primeros modelos.

### Matriz de correlación de Pearson de 14 variables

In [ ]:
mydata.corr(method="pearson")

CRIM        ZN     INDUS  ...         B     LSTAT      MEDV
CRIM     1.000000 -0.200469  0.406583  ... -0.385064  0.455621 -0.388305
ZN      -0.200469  1.000000 -0.533828  ...  0.175520 -0.412995  0.360445
INDUS    0.406583 -0.533828  1.000000  ... -0.356977  0.603800 -0.483725
CHAS    -0.055892 -0.042697  0.062938  ...  0.048788 -0.053929  0.175260
NOX      0.420972 -0.516604  0.763651  ... -0.380051  0.590879 -0.427321
RM      -0.219247  0.311991 -0.391676  ...  0.128069 -0.613808  0.695360
AGE      0.352734 -0.569537  0.644779  ... -0.273534  0.602339 -0.376955
DIS     -0.379670  0.664408 -0.708027  ...  0.291512 -0.496996  0.249929
RAD      0.625505 -0.311948  0.595129  ... -0.444413  0.488676 -0.381626
TAX      0.582764 -0.314563  0.720760  ... -0.441808  0.543993 -0.468536
PTRATIO  0.289946 -0.391679  0.383248  ... -0.177383  0.374044 -0.507787
B       -0.385064  0.175520 -0.356977  ...  1.000000 -0.366087  0.333461
LSTAT    0.455621 -0.412995  0.603800  ... -0.366087  1.000000 -0.737663
MEDV    -0.388305  0.360445 -0.483725  ...  0.333461 -0.737663  1.000000

[14 rows x 14 columns]

Los siguientes pares de variables obtuvieron mas de 0.7
1. INDUS vs TAX, DIS, NOX
1. NOX vs DIS, AGE, INDUS
1. AGE vs NOX, DIS
1. DIS vs INDUS, NOX, AGE
1. RAD vs TAX
1. TAX vs INDUS, RAD
1. LSTAT vs MEDV
1. MEDV vs LSTAT

Una vez que se tienen todas las parejas de variables independientes con coeficientes de correlación de Pearson mayores o iguales en valor absoluto a 0.7, se deben descartar una de ellas en cada pareja (usualmente la que tenga el coeficiente de pearson más bajo con la variable dependiente). Con las variables independientes que quedan, se genera el tercer modelo de regresión lineal múltiple con la variable dependiente MEDV.

Las columnas seleccionadas para el tercer modelo seran:
1. INDUS
1. NOX
1. AGE
1. TAX
1. LSTAT

---

### Tercer modelo RLM

In [ ]:
# Dejamos solo las columnas
x3 = x.loc[:, ["INDUS", "NOX", "AGE", "TAX", "LSTAT"]]
print(x3.shape)
print(x3.head())

xones3 = sm.add_constant(x3)  # Agregamos una columna de 1s

model3 = sm.OLS(y, xones3).fit()
print(model3.summary())

(506, 5)
   INDUS    NOX   AGE  TAX  LSTAT
0   2.31  0.538  65.2  296   4.98
1   7.07  0.469  78.9  242   9.14
2   7.07  0.469  61.1  242   4.03
3   2.18  0.458  45.8  222   2.94
4   2.18  0.458  54.2  222   5.33
                            OLS Regression Results                            
Dep. Variable:                   MEDV   R-squared:                       0.565
Model:                            OLS   Adj. R-squared:                  0.561
Method:                 Least Squares   F-statistic:                     129.8
Date:                Tue, 25 Aug 2020   Prob (F-statistic):           5.57e-88
Time:                        22:48:31   Log-Likelihood:                -1629.7
No. Observations:                 506   AIC:                             3271.
Df Residuals:                     500   BIC:                             3297.
Df Model:                           5                                         
Covariance Type:            nonrobust                                       

---

Se observa que el valor $\bar{R^2}=0.561$ es un tanto menor que el modelo con más variables, por el valor P podemos observar que en general los valores tienden más a 0 que indican que se están obteniendo los mejores parámetros que se pueden obtener con este set de entrenamiento.

Sin embargo, el valor P de NOX se disparó a 0.537, lo que parece indicar que dada la combinación de columnas, NOX no aporta mucho a la determinación de la variable dependiente. Pero el hecho de que NOX tuviera un valor P = 0 en el modelo con todas las variables pudiera sugerir más bien que la significancia de NOX depende de la presencia de otras variables independientes con las que tenga una fuerte correlación, como DIS.

Partiendo del supuesto anterior, podríamos sugerir un modelo con otro set de columnas que reemplace por ejemplo AGE por DIS, dado que AGE no aportaba mucho en nuestro modelo original y DIS posee una fuerte relación con NOX en la matriz de correlación. Podemos remover INDUS por la misma razón, TAX, NOX y DIS tenían fuerte relación con INDUS, así que esos pueden quedarse. Podríamos agregar RAD dada la fuerte relación con TAX.


In [ ]:
x3_2 = x.loc[:, ["NOX", "DIS", "TAX", "LSTAT", "RAD"]]

model3_2 = sm.OLS(y, sm.add_constant(x3_2)).fit()
print(model3_2.summary())

                            OLS Regression Results                            
Dep. Variable:                   MEDV   R-squared:                       0.594
Model:                            OLS   Adj. R-squared:                  0.590
Method:                 Least Squares   F-statistic:                     146.1
Date:                Tue, 25 Aug 2020   Prob (F-statistic):           2.32e-95
Time:                        23:01:18   Log-Likelihood:                -1612.4
No. Observations:                 506   AIC:                             3237.
Df Residuals:                     500   BIC:                             3262.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         49.3237      2.684     18.379      0.0

---

Podemos así ver que hay una mejora $\bar{R^2}=0.590$, el valor P de NOX se reduce considerablemente y en general el resto de los valores P se ven mucho mejor en comparación, lo que parece aceptar nuestro supuesto sobre la significancia de las variables y que el aporte que pudieran dar INDUS y AGE están mejor codificadas en las variables con las que tienen fuerte correlación.

##v. Con base a la matriz de correlación con las 14 variables identifica ahora aquellas variables independientes que tengan una correlación de pearson mayor o igual a 0.5 con respecto a la variable dependiente MEDV.
Obtener un cuarto modelo RLM y compararlo con los tres anteriores.

Se identifican aquellas variables independientes que tienen un coeficiente de Pearson mayor o igual a 0.5 en valor absoluto, con la variable dependiente. El cuarto modelo de RLM se genera usuando solamente las variables independientes que cumplieron dicho criterio y donde la variable dependiente sigue siendo MEDV.

Las variables a usar serán:
1. RM
1. PTRATIO
1.LTSTAT

---

### Cuarto modelo RLM


In [ ]:
# Dejamos solo las columnas RM, PTRATIO, LTSTAT
x4 = x.loc[:, ["RM", "PTRATIO", "LSTAT"]]
print(x4.shape)
print(x4.head())

xones4 = sm.add_constant(x4)  # Agregamos una columna de 1s
print(xones4.head())

model4 = sm.OLS(y, xones4).fit()
print(model4.summary())

(506, 3)
      RM  PTRATIO  LSTAT
0  6.575     15.3   4.98
1  6.421     17.8   9.14
2  7.185     17.8   4.03
3  6.998     18.7   2.94
4  7.147     18.7   5.33
   const     RM  PTRATIO  LSTAT
0    1.0  6.575     15.3   4.98
1    1.0  6.421     17.8   9.14
2    1.0  7.185     17.8   4.03
3    1.0  6.998     18.7   2.94
4    1.0  7.147     18.7   5.33
                            OLS Regression Results                            
Dep. Variable:                   MEDV   R-squared:                       0.679
Model:                            OLS   Adj. R-squared:                  0.677
Method:                 Least Squares   F-statistic:                     353.3
Date:                Mon, 24 Aug 2020   Prob (F-statistic):          2.69e-123
Time:                        22:45:09   Log-Likelihood:                -1553.0
No. Observations:                 506   AIC:                             3114.
Df Residuals:                     502   BIC:                             3131.
Df Model:        

---

Se observa que el valor $\bar{R^2}=0.677$ que es mejor que el modelo con aquellas
variables con un valor de coeficiente Pearson mayor a 0.7 además, por el valor P
podemos observar que se están obteniendo los mejores parámetros para este set de
entrenamiento.

El hecho que estemos usando menos variables que obtienen mejor valor $\bar{R^2}$ es evidencia directa de cómo RM, PTRATIO y LSTAT tienen una mayor correlación con la variable dependiente MEDV.

Sin embargo, y como pudimos ver con el segundo modelo, hay información valiosa en aquellas variables con coeficiente Pearson menor a 0.7, y esto parece ser comprobado en el primer modelo que incluía las 11 variables donde se consiguió un $\bar{R^2}=0.735$ que sigue siendo el mejor modelo.

##vi. Usando únicamente la matriz 𝑋 con las 3 variables independientes del inciso anterior y la variable de salida MDEV, obtener la matriz de Penrose

La matriz Penrose $X^+ = (X^TX)^{-1}X^T$

Posteriormente obtener los coeficientes 𝛽 del modelo RLM resultante mediante la expresión: $𝛽 = 𝑋^+𝑌$. Compáralo con el inciso anterior y escribe tus conclusiones.

In [ ]:
xtx = xones4.T.dot(xones4)
inverse = np.linalg.inv(xtx)
pearson_matrix = inverse.dot(xones4.T)

In [ ]:
beta_params = pearson_matrix.dot(y)
print(beta_params)

[[18.56711151]
 [ 4.51542094]
 [-0.93072256]
 [-0.57180569]]


---

Se obtuvieron los siguientes resultados para los parámetros de cada una de las variables:

Variable | OLS Params | Linear Algebra 
:---|:---:|:---:
const|18.5671|18.56711151
RM|4.5154|4.51542094
PTRATIO|-0.9307|-0.93072256
LSTAT|-0.5718|-0.57180569

El resultado de los parámetros es el mismo por OLS y por álgebra lineal.

##vii. Finalmente escribe tus conclusiones de los resultados obtenidos.
---

Los resultados obtenidos en todas las permutaciones variaron mucho, sin embargo, el mejor obtenido es aquel donde se obtuvieron $\bar{R^2} = 0.735$ usando 11 variables independientes (todas salvo INDUS y AGE). Esto se puede verificar con el bajo valor P para todos los coeficientes, lo que sugiere que son los mejores parámetros que se pueden conseguir.

En particular, podemos explicar que gran parte de ese 0.735 se debe en mayor medida a las variables RM, PTRATIO y LSTAT cuyo modelo alcanza $\bar{R^2}=0.677$, y la otra proporción se explica con la combinación de las restantes 8 variables independientes.

De este ejercicio podemos concluir que analizar el coeficiente de correlación entre variables independientes permite determinar qué variables deben ser descartadas o incluso cuáles pueden codificar relaciones implícitamente de manera que una de ellas se pueda eliminar sin que esto afecte el rendimiento del modelo.